tests/dragonfly/conftest.py

```diff
@@ -45,62 +39,64 @@ def test_env(tmp_dir: Path):
     return env
 
 
-@pytest.fixture(scope="class", params=[[]])
-def df_server(request, tmp_dir: Path, test_env):
-    """ Starts a single DragonflyDB process, runs once per test class. """
-    print(f"Starting DragonflyDB [{DRAGONFLY_PATH}]")
-    arguments = [arg.format(**test_env) for arg in request.param]
-    dfly_proc = subprocess.Popen([DRAGONFLY_PATH, *arguments],
-                                 env=test_env, cwd=str(tmp_dir))
-    time.sleep(0.3)
-    return_code = dfly_proc.poll()
-    if return_code is not None:
-        dfly_proc.terminate()
-        pytest.exit(f"Failed to start DragonflyDB [{DRAGONFLY_PATH}]")
-
-    yield
-
-    print(f"Terminating DragonflyDB process [{dfly_proc.pid}]")
-    try:
-        dfly_proc.terminate()
-        outs, errs = dfly_proc.communicate(timeout=15)
-    except subprocess.TimeoutExpired:
-        print("Unable to terminate DragonflyDB gracefully, it was killed")
-        outs, errs = dfly_proc.communicate()
-        print(outs)
-        print(errs)
+@pytest.fixture(scope="session", params=[{}])
+def df_factory(request, tmp_dir, test_env) -> DflyInstanceFactory:
+    """
+    Create an instance factory with supplied params.
+    """
+    scripts_dir = os.path.dirname(os.path.abspath(__file__))
+    path = os.environ.get("DRAGONFLY_PATH", os.path.join(
+        scripts_dir, '../../build-dbg/dragonfly'))
 
+    args = request.param if request.param else {}
+    return DflyInstanceFactory(test_env, tmp_dir, path=path, args=args)
 
-@pytest.fixture(scope="function")
-def connection(df_server):
-    return redis.Connection()
 
+@pytest.fixture(scope="session")
+def df_server(df_factory: DflyInstanceFactory) -> DflyInstance:
+    """
+    Start the default Dragonfly server that will be used for the default pools
+    and clients.
+    """
+    instance = df_factory.create()
+    instance.start()
+    yield instance
+    instance.stop()
```

Comment:
This fixture can be session level, because the previous instance is discarded anyways when a new one is created for fresh params

Reply:
All other fixtures as well, except those that depend on the function-scoped event loop

Reply:
The problem with session level is that you then holding states between tests. So if one test is running something like "set foo bar", and other test assumes that this foo do not exists before the test, now the test may fail (depending on the order you run it). I think it best not to hold an instance running across the tests

Reply:
@boazsade  the original thought was - to save time. I've seen pytests suits that grew to hundreds hours per run when most tests did not require a clean environment. I think for most cases "flushall" should suffice. We were also grouping tests by the running profile defined by dragonfly arguments.

Reply:
This is why the `client` and `async_client` are class scope and run flushall on creation. There should be no issues with keys. But I'm not sure about other things like leftover pub-sub subscribers... There connection_test currently doesn't close them. So lets maybe really switch back to class scope

Reply:
async_pool fixture is running at the function level, so when the function under test is done, it would clean the connection, which do unsbscribe and close the publisher as well. I don't think you need to do any changes there.

Reply:
I guess I just fix the connection_test so that all clients close